OpenRA.Mods.Common/Traits/Attack/AttackDetonates.cs

```diff
@@ -0,0 +1,140 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2016 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Orders;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Fires it's armament after a chargeup.")]
+	public class AttackDetonatesInfo : ITraitInfo
+	{
+		[Desc("Measured in ticks.")]
+		public readonly int ChargeDelay = 0;
+		public readonly string ChargeSound = null;
+
+		[Desc("Measured in ticks.")]
+		public readonly int DetonationDelay = 0;
+		[WeaponReference] public readonly string DetonationWeapon = null;
```

Comment:
I think this will want a `[FieldLoader.Require]`. (Or you need to check for `null` later on, line 67).